### PR TITLE
Improve paint/repaint behavior

### PR DESCRIPTION
**Pull request by bvaughn**
Makes several small changes to graphics related code to improve interruption and better handle caching edge cases. I would appreciate a very close review of `updateGraphics`. It _looks_ like it could be simplified but I don't think it can without missing some cases.

I wrote a new, more low-level repaint test (`repaint-06`) but it currently fails in many places due to a suspected RUN issue which Klochek is looking into (RUN-3397). For now, I have _disabled_ that test and filed a follow up issue (FE-2363).

### CODE REVIEW

File: packages/protocol/RepaintGraphicsCache.ts [hunk 1]
```diff
@@ -11,6 +12,20 @@ export const RepaintGraphicsCache: Cache<
...
+    const result = await replayClient.repaintGraphics(pauseId);
+
+    let { description, screenShot } = result;
+
+    // The backend won't return a screenshot for a given hash more than once; see FE-2357
+    if (screenShot) {
+      paintHashCache.cacheValue(screenShot, description.hash);
+    } else {
+      screenShot = paintHashCache.getValueIfCached(description.hash);
+    }
```
Comment (bvaughn) [inline packages/protocol/RepaintGraphicsCache.ts:24]: [`DOM.repaintGraphics` API docs](https://static.replay.io/protocol/tot/DOM/#method-repaintGraphics) say this about the `screenShot` value:
> Contents of the screen shot. If an identical screen shot was previously returned, this will be omitted.

File: packages/replay-next/src/suspense/ScreenshotCache.ts [hunk 2]
```diff
@@ -10,5 +10,18 @@ export const screenshotCache = createCache<
   config: { immutable: true },
   debugLabel: "ScreenshotCache",
   getKey: ([client, point, paintHash]) => paintHash,
-  load: ([client, point]) => client.getScreenshot(point),
+  load: async ([client, point]) => {
+    const screenShot = await client.getScreenshot(point);
+
+    paintHashCache.cacheValue(screenShot, screenShot.hash);
+
+    return screenShot;
```
Comment (bvaughn) [inline packages/replay-next/src/suspense/ScreenshotCache.ts:18]: I'm not sure if this is prone to the same case mentioned in the repaint cache. Seems like it could be, depending on how RUN keys screenshots (by point or by hash)

File: src/ui/components/Video/imperative/MutableGraphicsState.tsx [hunk 3]
```diff
@@ -48,25 +50,33 @@ export const state = createState<State>({
...
     didResize?: boolean;
     executionPoint: ExecutionPoint | null;
     screenShot: ScreenShot | null;
+    screenShotType: ScreenShotType | null;
     status: Status;
     time: number;
   }> = {}
 ) {
+  const localLock = new Object();
+  lock = localLock;
```
Comment (bvaughn) [inline src/ui/components/Video/imperative/MutableGraphicsState.tsx:71]: There was a potential race here before when e.g. scrubbing the timeline.

File: src/ui/components/Video/imperative/updateGraphics.ts [hunk 4]
```diff
@@ -31,27 +31,22 @@ export async function updateGraphics({
...
+  const isBeforeFirstCachedPaint = !paintPoint || !paintPoint.paintHash;
+  if (!isBeforeFirstCachedPaint) {
+    promises.push(
+      fetchPaintContents({
+        abortSignal,
+        containerElement,
+        replayClient,
+        time,
+      })
+    );
```
Comment (bvaughn) [inline src/ui/components/Video/imperative/updateGraphics.ts:47]: Even if we're before the first paint, we can still request a repaint.

**Comment by jasonLaster** [comment]
Love seeing the new repaint test!

File: packages/e2e-tests/tests/repaint-06.test.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      await stepOver(page);
+      await stepOver(page);
+
+      await waitForRepaintGraphicsToLoad(page);
+
+      graphicsDataUrl = await getGraphicsDataUrl(page);
+      expect(
+        graphicsDataUrl,
+        `Expected graphics before "${string}" not to match previous after graphics`
+      ).not.toEqual(previousGraphicsDataUrl);
```
Comment (hbenl) [inline packages/e2e-tests/tests/repaint-06.test.ts:62]: I don't think this is correct: why should the screenshot change in between updates?

File: src/ui/components/Video/imperative/MutableGraphicsState.tsx [hunk 6]
```diff
@@ -86,6 +100,10 @@ export async function updateState(
       imageHeight: naturalHeight,
       imageWidth: naturalWidth,
     });
+    if (lock !== localLock) {
+      return;
```
Comment (hbenl) [inline src/ui/components/Video/imperative/MutableGraphicsState.tsx:104]: Nit: `fitImageToContainer()` isn't async so we can remove the `await` and this check.

**Review by hbenl — APPROVED** [review]

File: packages/e2e-tests/tests/repaint-06.test.ts (hunk 5 above)
Comment (bvaughn) [reply packages/e2e-tests/tests/repaint-06.test.ts:62]: Yeah, you're right about that.
